src/VisualStudio/Core/Def/Implementation/CodeLens/RemoteCodeLensReferencesService.cs

```diff
@@ -104,11 +97,135 @@ internal sealed class RemoteCodeLensReferencesService : ICodeLensReferencesServi
+
+                var document = solution.GetDocument(referencedDocumentId);
+
+                var spanMapper = document?.Services.GetService<ISpanMappingService>();
+                if (spanMapper == null)
+                {
+                    // for normal document, just add one as they are
+                    list.Add(descriptor);
+                    continue;
+                }
+
+                var span = new TextSpan(descriptor.SpanStart, descriptor.SpanLength);
+                var results = await spanMapper.MapSpansAsync(document, SpecializedCollections.SingletonEnumerable(span), cancellationToken).ConfigureAwait(false);
+
+                // external component violated contracts. the mapper should preserve input order/count. 
+                // since we gave in 1 span, it should return 1 span back
+                Contract.ThrowIfTrue(results.IsDefaultOrEmpty);
+
+                var result = results[0];
+                if (result.IsDefault)
+                {
+                    // it is allowed for mapper to return default 
+                    // if it can't map the given span to any usable span
+                    continue;
+                }
+
+                var excerpter = document.Services.GetService<IDocumentExcerptService>();
+                var referenceExcerpt = await excerpter.TryExcerptAsync(document, span, ExcerptMode.SingleLine, cancellationToken).ConfigureAwait(false);
+                var tooltipExcerpt = await excerpter.TryExcerptAsync(document, span, ExcerptMode.Tooltip, cancellationToken).ConfigureAwait(false);
+
+                var reference = GetReferenceInfo(referenceExcerpt, descriptor);
+                var referenceTexts = GetReferenceTexts(referenceExcerpt, tooltipExcerpt, descriptor);
+
+                list.Add(new ReferenceLocationDescriptor(
+                    descriptor.LongDescription,
+                    descriptor.Language,
+                    descriptor.Glyph,
+                    result.Span.Start,
+                    result.Span.Length,
+                    result.LinePositionSpan.Start.Line,
+                    result.LinePositionSpan.Start.Character,
+                    descriptor.ProjectGuid,
+                    descriptor.DocumentGuid,
+                    result.FilePath,
+                    reference.text,
+                    reference.start,
+                    reference.length,
+                    referenceTexts.before1,
+                    referenceTexts.before2,
+                    referenceTexts.after1,
+                    referenceTexts.after2));
+            }
+
+            return list;
+        }
+
+        private static (string text, int start, int length) GetReferenceInfo(ExcerptResult? reference, ReferenceLocationDescriptor descriptor)
+        {
+            if (reference.HasValue)
+            {
+                return (reference.Value.Content.ToString().TrimEnd(),
```

Comment:
why are we trimming?  code needs comments to explain what's going on.

Reply:
keeping existing behavior.